### PR TITLE
feat(webpack-cli): infrastructure logging

### PR DESCRIPTION
**Pull request by piecyk**
**What kind of change does this PR introduce?**

Using InfrastructureLogger for logging output from webpack compiler

**Did you add tests for your changes?**

I plan to do it

**If relevant, did you update the documentation?**

No

**Summary**

Related #2019

**Does this PR introduce a breaking change?**

Yes

**Other information**

### CODE REVIEW

**Comment by piecyk** [comment]
As mention before webpack infrastructure logging outputs to `stderr` that affect's lot's of test.

File: packages/webpack-cli/__tests__/info.test.js [hunk 1]
```diff
@@ -3,28 +3,26 @@ const path = require('path');
...
     it('should work with flags', () => {
-        const { stdout, stderr } = spawnSync(path.resolve(__dirname, '../bin/cli.js'), ['info', '--output=json'], {
+        const { stderr } = spawnSync(path.resolve(__dirname, '../bin/cli.js'), ['info', '--output=json'], {
             cwd: path.resolve(__dirname),
             reject: false,
         });
 
         const testJSON = () => {
-            const output = JSON.parse(stdout);
+            const output = JSON.parse(stderr);
```
Comment (evilebottnawi) [inline packages/webpack-cli/__tests__/info.test.js:22]: Breaking change

File: packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js [hunk 2]
```diff
@@ -1,62 +1,143 @@
...
-const logger = require('../utils/logger');
+const { getStatsOptions } = require('../utils/stats-options');
+const { PluginName } = require('../utils/name');
 
-const PluginName = 'webpack-cli';
-
-class WebpackCLIPlugin {
-    constructor(options) {
-        this.options = options;
+class WebpackCLIRuntimePlugin {
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:6]: No need to rename, keep WebpackCLIPLugin

File: packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js [hunk 3]
```diff
@@ -1,62 +1,143 @@
...
+    constructor() {
+        this.isWatch = false;
     }
 
     async apply(compiler) {
-        const compilers = compiler.compilers || [compiler];
+        const isWebpack5 = !!compiler.webpack;
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:12]: Boolean(compiler.webpack)

File: packages/webpack-cli/lib/utils/logger.js [hunk 4]
```diff
@@ -1,11 +1,25 @@
...
+    log: write(),
+
+    /**
+     * @deprecated use info
+     */
+    success: write(undefined, green),
+    /**
+     * @deprecated use info
+     */
+    raw: write(''),
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/utils/logger.js:24]: Don't touch it here

File: packages/webpack-cli/lib/utils/logger.js [hunk 5]
```diff
@@ -1,11 +1,25 @@
 const util = require('util');
 const { red, cyan, yellow, green } = require('colorette');
 
+const write = (prefix = '[webpack-cli] ', color = (val) => val) => {
+    return (...args) => {
+        const str = prefix + color(util.format(...args));
+        process.stderr.write(str + '\n');
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/utils/logger.js:7]: No, it is real bad idea

**Review by evilebottnawi — CHANGES_REQUESTED** [review]
Sorry, it is big breaking change, do not touch stdout and stderr, it was big problems for previously versions and we fix it

**Comment by piecyk** [comment]
@evilebottnawi weback InfrastructureLogger is writing to `stderr` https://github.com/webpack/webpack/blob/master/lib/node/nodeConsole.js#L48

Aligned the utils/logger for simpler refactor test, and not go one by one 🙈  Let's revert logger to console and fix the test, just ping me if that is something we want as it's not a small thing.

File: packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js (hunk 3 above)
Comment (piecyk) [reply packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:12]: Why it's better ?

File: packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js (hunk 2 above)
Comment (piecyk) [reply packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:6]: Yep, just was playing around with some ideas. Would be nice if `webpack-dev-middleware` and `webpack-cli` output messages in same format. Basic sharing the plugin https://github.com/webpack/webpack-dev-middleware/blob/master/src/utils/setupHooks.js

**Comment by evilebottnawi** [comment]
hm, I need dicussion with sokra

**Comment by evilebottnawi** [comment]
Anyway we can't change it here, because it is breaking changte

**Comment by piecyk** [comment]
Looks like reverting logger was not that bad. Still needed to disable few test, like with `--bail` is something off...

But yeah, also proposing some changes in what we log, now we show 

`Compilation finished` at end, and simliar information we show in stats, 
`webpack 5.4.0 compiled successfully in 610 ms`. Imho better to use this rather finished

something like
 
v4
```
$ webpack --watch --progress
<i> [webpack-cli] compilation starting...
<i> [webpack-cli] output:
<i> Hash: 01af0357780e7ff104a9
<i> Built at: 11/06/2020 12:41:15 PM
<i>   Asset      Size  Chunks             Chunk Names
<i> main.js  4.01 KiB    main  [emitted]  main
<i> Entrypoint main = main.js
<i> [./src/index.ts] 53 bytes {main} [built]
<i> [webpack-cli] webpack 4.44.2 compiled in 1871ms
<i> [webpack-cli] watching files for updates...
```

v5
```
$ webpack --watch --progress
<i> [webpack-cli] compilation starting...
<i> [webpack-cli] output:
<i> asset main.js 3.79 KiB [emitted] (name: main)
<i> runtime modules 668 bytes 3 modules
<i> ./src/index.ts 53 bytes [built] [code generated]
<i> [webpack-cli] webpack 5.4.0 compiled in 632 ms
<i> [webpack-cli] watching files for updates...
```

**Comment by evilebottnawi** [comment]
Yes, we should remove `Compilation finished`

**Comment by evilebottnawi** [comment]
`<i> [webpack-cli] output:` is unnecessary and looks like spam

**Comment by evilebottnawi** [comment]
And yes, you will right we should - All diagnostic output goes to stderr. only the stats output goes to stdout

File: packages/webpack-cli/lib/utils/stats-options.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+            options = webpack.Stats.presetToOptions(options);
+        }
+    }
+
+    options.colors = typeof options.colors !== 'undefined' ? options.colors : coloretteOptions.enabled;
+
+    return options;
+};
+
+module.exports = { getStatsOptions };
```
Comment (evilebottnawi) [inline packages/webpack-cli/lib/utils/stats-options.js:26]: No utils relates to complation, use methods classes

**Review by evilebottnawi — COMMENTED** [review]
I think we should avoid do here a lot of changes, let's separate them, I can't remove and merge this all big changes, sorry, let's focus on migrate infrastructure logger

**Comment by piecyk** [comment]
> I think we should avoid do here a lot of changes, let's separate them, I can't remove and merge this all big changes, sorry, let's focus on migrate infrastructure logger

Yes i know. Basic problem is that migrating to infrastructure logger is bigger one as it touch a lot of testes as we switch from.

> And yes, you will right we should - All diagnostic output goes to stderr. only the stats output goes to stdout

So basic when logging stats we should not use infrastructure logger ?

File: packages/webpack-cli/lib/utils/stats-options.js (hunk 6 above)
Comment (piecyk) [reply packages/webpack-cli/lib/utils/stats-options.js:26]: This was execrated to utils as, moved stats display to WebpackCLIPlugin, but wanted to keep `json output` in WebpackCLI compiler callback.

**Comment by evilebottnawi** [comment]
Yes, for stats we should use `stdout` and do not use infrastructure logic

**Comment by evilebottnawi** [comment]
> Yes i know. Basic problem is that migrating to infrastructure logger is bigger one as it touch a lot of testes as we switch from.

So let's separate it

**Comment by piecyk** [comment]
> So let's separate it

ok current plan 

* [x] revert utils/logger changes. 
* [x] use infrastructure logger in context of webpack
* [ ] stats should log to stdout 
* [ ] fix test

File: packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js [hunk 7]
```diff
@@ -1,63 +1,120 @@
...
+                        });
 
-                    const isProfile = this.options.progress === 'profile';
+                        return `${name ? name + `(${version})` : version} compiled in ${time}`;
+                    }
+                })();
 
-                    new ProgressPlugin({ profile: isProfile }).apply(compiler);
+                if (childStats.hasErrors()) {
+                    logger.error(normalizedStatus);
```
Comment (snitin315) [inline packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:49]: We should set exitCode = 1 for errors.

Comment (piecyk) [reply packages/webpack-cli/lib/plugins/WebpackCLIPlugin.js:49]: Exit code is set in callback in webpack-cli.js this wasn't changed https://github.com/webpack/webpack-cli/pull/2036/files#diff-7c711e2a4ef4b16d2cbbe1658495723402f9692da654cb6fa959127f3dc779f2R357-R359

Wondering if we should reset it if next complication is has no errors 🤔

**Comment by piecyk** [comment]
Suppressed by #2144

**Pull request closed without merging**